guava/src/com/google/common/collect/ImmutableSet.java

```diff
@@ -722,20 +739,92 @@ static int maxRunBeforeFallback(int tableSize) {
     RegularSetBuilderImpl(RegularSetBuilderImpl<E> toCopy) {
       super(toCopy);
       this.hashTable = Arrays.copyOf(toCopy.hashTable, toCopy.hashTable.length);
+      this.numElementsInHashTable = toCopy.numElementsInHashTable;
       this.maxRunBeforeFallback = toCopy.maxRunBeforeFallback;
       this.expandTableThreshold = toCopy.expandTableThreshold;
       this.hashCode = toCopy.hashCode;
+
+      if (toCopy.overflowTables != null) {
+        int size = toCopy.overflowTables.size();
+        this.overflowTables = new ArrayList<>(size);
+        for (int i = 0; i < size; i++) {
+          Object[] table = toCopy.overflowTables.get(i);
+          this.overflowTables.add(Arrays.copyOf(table, table.length));
+        }
+      }
     }
 
+    @SuppressWarnings("unchecked")
     void ensureTableCapacity(int minCapacity) {
       if (minCapacity > expandTableThreshold && hashTable.length < MAX_TABLE_SIZE) {
-        int newTableSize = hashTable.length * 2;
-        hashTable = rebuildHashTable(newTableSize, dedupedElements, distinct);
+        /**
+         * we have about {@code (a_0 sum_{k=1 to n+1} of 2^k)} elements total in our table, assuming
+         * we fill each of the overflow tables fully, which is {@code a_0 * (2^{n+2} - 2) <=
+         * hashtable.size * 2} by geometric series.
+         *
+         * <p>We always allocate enough to fill all elements in our table to needing to rehash
+         * frequently.
+         */
+        int newTableSize = chooseTableSize(hashTable.length * 2);
+        if (overflowTables == null) {
+          overflowTables = new ArrayList<>();
+        }
+        overflowTables.add(dedupedElements);
+        distinct = 0;
+        dedupedElements = (E[]) new Object[expandTableThreshold];
+
+        numElementsInHashTable = 0;
+        hashTable = new Object[newTableSize];
+
+        hashCode = 0;
         maxRunBeforeFallback = maxRunBeforeFallback(newTableSize);
         expandTableThreshold = (int) (DESIRED_LOAD_FACTOR * newTableSize);
       }
     }
 
+    @SuppressWarnings("unchecked")
+    void flattenOverflowTables() {
+      if (overflowTables == null) {
+        return;
+      }
+
+      // ensure capacity first to avoid resizing and inserting into a high load table as much as
+      // possible
+      ensureTableCapacity(hashTable.length * 2);
+
+      ArrayList<Object[]> overflows = overflowTables;
+      overflowTables = null;
+      // now we rehash.
+      for (int i = 0; i < overflows.size(); i++) {
+        Object[] table = overflows.get(i);
+        for (int j = 0; j < table.length && table[j] != null; j++) {
```

Comment:
I'm concerned about the semantics of this, unfortunately: it seems like it'd destroy insertion order, because elements from overflowTables are getting added *after* elements from the current chunk, when they were added to the builder before.  I'm not sure I see a good way around that. =(

Reply:
Yeah this was a bug I found. I'm closing this pull request due to some new findings that I'm going to post on the issue.